doc/man/man4/itos.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/itos.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH ITOS 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B isos, itos
\- ISO Transport over SCTP (ISOS/ITOS)
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <stropts.h>
.HP
.IB sctp_stream " = open( /dev/sctp_t ", O_RDWR);
.HP
.BI ioctl( sctp_stream ", I_PUSH, " \(lqisos\(rq );
.sp
.HP
.IB sctp_stream " = open( /dev/sctp_t ", O_RDWR);
.HP
.BI ioctl( sctp_stream ", I_PUSH, " \(lqitos\(rq );
.\"
.\"
.SH DESCRIPTION
.PP
.B ISOS
is an implementation of ISO Transport over SCTP (ISOS) similar to that
described for TCP in RFC 983,
.[
rfc983
.]
and superseded by RFC 1006.
.[
rfc1006
.]
.B ITOS 
is an implementation of ISO Transport over SCTP (ITOS) similar to that
described for TCP in RFC 2126.
.[
rfc2126
.]
In accordance with the naming convention for the TCP versions of the
protocols, the SCTP version
.I ISOS
refers to the RFC 1006
.[
rfc1006
.]
flavor for carrying TP0 only, and the
.I ITOS
refers to the RFC 2126
.[
rfc2126
.]
flavor for carrying TP2 as well.
.PP
Regardless of the naming differences, these RFCs describe the use of IS 8073
.[
ISO/IEC 8073
.]
ISO Transport Class 0 over TCP/IP (ISOT/ITOT) by using TCP/IP as a
Connection-oriented Network service Provider instead of ISO CONS.
.[
x213
.]
They describe the adaptation of TCP to an ISO CONS
.[
x213
.]
network service provider through the use of a 4 byte header used to delimit
NPDUs inside of the TCP stream.
RFC 983
.[
rfc983
.]
and RFC 1006
.[
rfc1006
.]
provided for ISO Transport Class 0 over TCP (or UDP under
protest) (ISOT) on port number 102.
RFC 1006
is a standard (STD 35).
.PP
RFC 2126
.[
rfc2126
.]
also provides for ISO Transport Class 0 over TCP (or UDP under protest) (ISOT)
on port number 102.  But, in addition to these services provided by RFC
1006/STD 35, RFC 2126 also provides for ISO Transport Class 2 over TCP (or UDP
under protest) (ITOT), using independent TCP/IP connections to port number
399, and also supports both IPv4 and IPv6.
.PP
In a similar way
.BR ISOS / ITOT
provide a mechanism for SCTP (Stream Control Transmission Protocol, RFC 2960
.[[
rfc2960
.]]).
Under SCTP, however, the use of the 4-byte header to delimit NPDUs is no
longer necesary, as SCTP preserves record boundaries.  Also, the TCP flavor
must open a separate TCP stream for each ISO CONS VC, whereas SCTP can support
multiple VCs on a single association.
Also, there is no reason why SCTP cannot suport the other transport protocol
classes such as TP1 TP3 and TP4.
.\"
.\"
.SH USAGE
.PP
Not impelmented, but here is a stab at explaining the design:
.PP
The following steps will be taken:
.IP 1. \w'0.\(em'u
A TP connection manager opens a connection management Stream on the TP
multiplexing driver and listens for requests for the establishment of network
connections.
.IP 2.
A TP user opens an upper Stream for ISO Transport Class 0 or ISO Transport
Class 2.  The TP user then attempts t listen on the transport endpoint or
connect the transport endpoint.  If a suitable (determined by address format,
TSAP, and ownership) network connection is not available, it requests a
network connection from the connection manager. (TP2 performs multiplexing, so
it is possible that many TP2 Streams could be opened and served by the same
network connection.)
.IP 3.
When the connection manager receives the request, it uses the address format,
TSAP and qos parameters to determin if an ISOS or ITOS network connection is
requested.
.IP 4.
If an ISOS or ITOS network connection is requested, the connection manager
opens an
.BR sctp (4),
.BR tcp (4),
or
.BR udp (4)
Stream to the requested address and connects the Stream to the appropriate
port number (102 for ISOS TP0 and 399 for ITOS TP2).
.IP 5.
The connection manager then pushes the ISOS or ITOS moudl on the Stream.  This
transforms the Stream into an ISO CONS
.[
x213
.]
provider and presents the
.BR npi (7)
service interface at the upper boundary.
.IP 6.
The connection manager then links the ISOS or ITOS Stream under the TP
multiplexing driver and configures the Stream with the TP driver in response
to the network connection request.
.IP 7.
TP then proceeds as normal, unaware that SCTP is providing the ISO CONS
.[
x213
.]
service.
.\"
.\"
.SH INTERFACE
.PP
The service interface for the ISOS and ITOS modules at the upper boundary
(write queue) is the Network Provider Interface (NPI),
.[
npi revision 2.0.0 draft 2
.]
.BR npi (7).
The service interace for the ISOS and ITOS modules at the lower boundary (read
queue) is the Transport Provider Interface (TPI),
.[
tpi revision 2.0.0 draft 2
.]
.BR tpi (7).
.\"
.\"
.SH ADDRESSES
.PP
Generic string addresses for ISOS and ITOS are described in RFC 1278.
.[
rfc1278
.]
.PP
OSI NSAP addresses for ISOS and ITOS are described for IPv4 in RFC 1277,
.[
rfc1277
.]
for IPv6 in RFC 1888
.[
rfc1888
.]
superseded by IPv4 and IPv6 in RFC 4548.
.[
rfc4548
.]
.\"
.\"
.SH OPTIONS
.PP
.B ISOS
and
.B ITOS
modules support the ISO CONS
options described in X.213
.[
x213
.]
and X.223.
.[
x223
.]
.\"
.\"
.SH PRIMITIVES
.PP
The
ISOS and ITOS modules interwork
.BR npi (7)
primitives from its upper interface (write queue) to 
.BR tpi (7)
primitives at its lower interface (read queue).
This section describes only the upper interface.  For a description of the
lower interface, see
.BR sctp (4).
.SS N_BIND_REQ\fR(7), \fPN_BIND_ACK\fR(7)
.\"
.\"
.SH IOCTLS
.PP
ISOS and ITOS do not support any input-output controls of their own.
.\"
.\"
.SH NOTICES
.PP
These modules are not implemented.
.\"
.\"
.SH IMPLEMENTATION
.PP
These modules are not implemented.
.\"
.\"
.\".SH FILES
.\".PP
.\"
.\"
.SH MODULES
.PP
.TP
.B isos
the ISOS module is a
.I STREAMS
module named
.RI \(lq isos \(rq.
.TP
.B itos
the ITOS module is a
.I STREAMS
module named
.RI \(lq itos \(rq.
.\"
.\"
.SH DEVICES
.PP
.TP
.B /dev/sctp_t
The
.B sctp (4)
device opened onto which ISOS or ITOS is pushed.
.\"
.\"
.SH "SEE ALSO"
.PP
.IR ISO (4).
.\"
.\"
.SH BUGS
.PP
These modules are not implemented and so cannot have any bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
This implementation of ISOS and ITOS is specific to
.BR "@PACKAGE_TITLE@" ,
but is compatible with the comparable
.BR isot (4)
and
.BR itot (4)
implementations for TCP/IP.
.\"
.\"
.SH CONFORMANCE
.PP
ISOS and ITOS have not yet been subsjected to standardization.
.\"
.\"
.SH HISTORY
.PP
.BR ISOS " and " ITOS
are new for
.BR "@PACKAGE_TITLE@" .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/modules/itos.c
/*****************************************************************************

 @(#) File: src/modules/itos.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/modules/itos.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 *  ITOS module
 */

#define _SVR4_SOURCE
#define _MPS_SOURCE
#define _SUN_SOURCE

#include <sys/os7/compat.h>

#define ITOS_DESCRIP	"ITOS STREAMS Module"
#define ITOS_EXTRA	"Part of the OpenSS7 OSI Stack for Linux Fast-STREAMS"
#define ITOS_REVISION	"OpenSS7 src/modules/itos.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define ITOS_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define ITOS_DEVICE	"Provides OpenSS7 ISO Transport over SCTP (ITOS)"
#define ITOS_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define ITOS_LICENSE	"GPL"
#define ITOS_BANNER	ITOS_DESCRIP	"\n" \
			ITOS_EXTRA	"\n" \
			ITOS_REVISION	"\n" \
			ITOS_COPYRIGHT	"\n" \
			ITOS_DEVICE	"\n" \
			ITOS_CONTACT	"\n"
#define ITOS_SPLASH	ITOS_DEVICE	" - " \
			ITOS_REVISION	"\n"

#ifdef LINUX
MODULE_AUTHOR(ITOS_CONTACT);
MODULE_DESCRIPTION(ITOS_DESCRIP);
MODULE_SUPPORTED_DEVICE(ITOS_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(ITOS_LICENSE);
#endif				/* MODULE_LICENSE */
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-itos");
#endif				/* MODULE_ALIAS */
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif				/* MODULE_VERSION */
#endif				/* LINUX */

#define ITOS_MOD_ID	CONFIG_STREAMS_ITOS_MODID
#define ITOS_MOD_NAME	CONFIG_STREAMS_ITOS_NAME

#ifndef ITOS_MOD_NAME
#define ITOS_MOD_NAME	"itos"
#endif				/* ITOS_MOD_NAME */

#ifndef ITOS_MOD_ID
#define ITOS_MOD_ID	0
#endif				/* ITOS_MOD_ID */

/*
 *  STREAMS DEFINITIONS
 *  -------------------------------------------------------------------------
 */

#define MOD_ID		ITOS_MOD_ID
#define MOD_NAME	ITOS_MOD_NAME
#ifdef MODULE
#define MOD_BANNER	ITOS_BANNER
#else				/* MODULE */
#define MOD_BANNER	ITOS_SPLASH
#endif				/* MODULE */

/*
 *  LINUX INITIALIZATION
 *  -------------------------------------------------------------------------
 */

#ifdef LINUX
unsigned short modid = MOD_ID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else				/* module_param */
module_param(modid, short, 0444);
#endif				/* module_param */
MODULE_PARM_DESC(modid, "Module ID for ITOS module.  (0 for allocation.)");

/** itosinit - initialize ITOS
  */
static __init int
itosinit(void)
{
	cmn_err(CE_NOTE, MOD_BANNER);
	return (0);
}

/** itosexit - terminate ITOS
  */
static __exit void
itosexit(void)
{
	return;
}

module_init(itosinit);
module_exit(itosexit);

#endif				/* LINUX */
